nomos-core/chain-defs/src/mantle/genesis_tx.rs

```diff
@@ -0,0 +1,391 @@
+use groth16::Fr;
+use poseidon2::Digest;
+use serde::Serialize;
+
+use crate::{
+    crypto::ZkHasher,
+    mantle::{
+        AuthenticatedMantleTx, MantleTx, SignedMantleTx, Transaction, TransactionHasher, TxHash,
+        gas::{Gas, GasConstants, GasCost},
+        ops::{
+            Op, OpProof,
+            channel::{ChannelId, MsgId, inscribe::InscriptionOp},
+        },
+    },
+    proofs::zksig::ZkSignatureProof,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+pub struct GenesisTx(SignedMantleTx);
+
+#[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+pub enum Error {
+    #[error("Genesis transaction must have gas price of zero")]
+    InvalidGenesisGasPrice,
+    #[error("Genesis transaction should not have any inputs")]
+    UnepectedInput,
+    #[error("Genesis block cannot contain this op: {0:?}")]
+    UnsupportedGenesisOp(Vec<Op>),
+    #[error("Expected exactly one inscription in genesis block")]
+    MissingInscription,
+    #[error("Invalid genesis inscription: {0:?}")]
+    InvalidInscription(Box<Op>),
+}
+
+impl GenesisTx {
+    pub fn from_tx(signed_tx: SignedMantleTx) -> Result<Self, Error> {
+        let mantle_tx = signed_tx.mantle_tx();
+
+        // Genesis transactions must have gas prices of zero
+        if mantle_tx.execution_gas_price != 0 || mantle_tx.storage_gas_price != 0 {
+            return Err(Error::InvalidGenesisGasPrice);
+        }
+
+        // Genesis transactions should not have any inputs
+        if !mantle_tx.ledger_tx.inputs.is_empty() {
+            return Err(Error::UnepectedInput);
+        }
+
+        // Genesis transactions must contain exactly one inscription as the first op
+        // and then may contain other SDP declarations
+        let mut ops = mantle_tx.ops.iter();
+        match ops.next() {
+            Some(Op::ChannelInscribe(op)) => validate_inscription_op(op)?,
+            _ => return Err(Error::MissingInscription),
+        }
+
+        let unsupported_ops = ops
+            .filter(|op| !matches!(op, Op::SDPDeclare(_)))
+            .cloned()
+            .collect::<Vec<_>>();
+        if !unsupported_ops.is_empty() {
+            return Err(Error::UnsupportedGenesisOp(unsupported_ops));
+        }
+
+        Ok(Self(signed_tx))
+    }
+}
+
+fn validate_inscription_op(inscription: &InscriptionOp) -> Result<(), Error> {
+    if inscription.parent != MsgId::root() {
+        return Err(Error::InvalidInscription(Box::new(Op::ChannelInscribe(
+            inscription.clone(),
+        ))));
+    }
+
+    if inscription.channel_id != ChannelId::from([0; 32]) {
+        return Err(Error::InvalidInscription(Box::new(Op::ChannelInscribe(
+            inscription.clone(),
+        ))));
+    }
+
+    if inscription.signer.as_bytes() != &[0; 32] {
+        return Err(Error::InvalidInscription(Box::new(Op::ChannelInscribe(
+            inscription.clone(),
+        ))));
+    }
+
+    Ok(())
+}
```

Comment:
two things,

1. This inscription is specifically an inscription to initialize cryptarchia. So we can make that concrete by renaming this to `validate_cryptarchia_inscription()`

2. We should validate that the inscription payload `json::decodes` into a struct like:

```rust
struct CryptarchiaGenesisParams {
    chain_id: String,
    genesis_time: DateTime,
    genesis_epoch_nonce: Nonce,
}
```

Reply:
discussed offline, cryptarchia params should probably have a canonical encoding. We'll revisit that in the spec